src/components/SignUpForm/index.test.js

```diff
@@ -0,0 +1,409 @@
+import "@testing-library/jest-dom/extend-expect";
+
+import SignUpForm from "./index";
+
+describe("SignUpForm", () => {
+  it("renders correctly and matches the snapshot", () => {
+    const tree = renderer.create(<SignUpForm />).toJSON();
+    expect(tree).toMatchSnapshot();
+  });
+
+  const mockSubmit = jest.fn(
+    (firstName, lastName, email, password, confirmPassword, acceptTerms) => {
+      return Promise.resolve({
+        firstName,
+        lastName,
+        email,
+        password,
+        confirmPassword,
+        acceptTerms,
+      });
+    }
+  );
+
+  beforeEach(() => {
+    render(<SignUpForm submit={mockSubmit} />);
+  });
+
+  it("should display required error when value is empty", async () => {
+    fireEvent.submit(screen.getByRole("button"));
+
+    expect(await screen.findAllByRole("alert")).toHaveLength(5);
+    expect(mockSubmit).not.toBeCalled();
+  });
+
+  it("should display matching error when email is invalid", async () => {
+    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
+      target: {
+        value: "name",
+      },
+    });
+
+    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
+      target: {
+        value: "surname",
+      },
+    });
+
+    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
+      target: {
+        value: "test",
+      },
+    });
+
+    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
+      target: {
+        value: "password",
+      },
+    });
+
+    fireEvent.input(screen.getByPlaceholderText(/repeat password/i), {
+      target: {
+        value: "password",
+      },
+    });
+
+    fireEvent.click(screen.getByRole("checkbox"));
+
+    fireEvent.submit(screen.getByRole("button"));
+
+    expect(await screen.findAllByRole("alert")).toHaveLength(1);
+    expect(mockSubmit).not.toBeCalled();
+    expect(screen.getByRole("textbox", { name: /firstName/i }).value).toBe(
+      "name"
+    );
+    expect(screen.getByRole("textbox", { name: /lastName/i }).value).toBe(
+      "surname"
+    );
+    expect(screen.getByRole("textbox", { name: /email/i }).value).toBe("test");
+    expect(screen.getByPlaceholderText(/^password/i).value).toBe("password");
+    expect(screen.getByPlaceholderText(/repeat password/i).value).toBe(
+      "password"
+    );
+    expect(screen.getByRole("checkbox").checked).toBe(true);
+  });
+
+  it("should display matching error when email domain is invalid", async () => {
+    fireEvent.input(screen.getByRole("textbox", { name: /firstName/i }), {
+      target: {
+        value: "name",
+      },
+    });
+
+    fireEvent.input(screen.getByRole("textbox", { name: /lastName/i }), {
+      target: {
+        value: "surname",
+      },
+    });
+
+    fireEvent.input(screen.getByRole("textbox", { name: /email/i }), {
+      target: {
+        value: "test@test.co",
+      },
+    });
+
+    fireEvent.input(screen.getByPlaceholderText(/^password/i), {
```

Comment:
why do you use name to get inputs above but placeholder text here? placeholder text is also more brittle

Reply:
this is the more direct and easy way to reach elements as I found. Unfortunately, the [queries](https://testing-library.com/docs/dom-testing-library/api-queries/#queries) that React testing library use are so few and depend in general on the content shown.